src/optimisation/Optimisation.jl

```diff
@@ -1,324 +1,327 @@
+    inner_context = estimator isa MAP ? DefaultContext() : LikelihoodContext()
+    ctx = OptimizationContext(inner_context)
+    log_density = OptimLogDensity(model, ctx)
+    return ModeEstimationProblem(
+        model, estimator, init_value, constraints, log_density, adtype, false
+    )
 end
 
+has_box_constraints(p::ModeEstimationProblem) = has_box_constraints(p.constraints)
+has_generic_constraints(p::ModeEstimationProblem) = has_generic_constraints(p.constraints)
 
+function default_solver(problem::ModeEstimationProblem)
+    return has_generic_constraints(problem.constraints) ? IPNewton() : LBFGS()
+end
 
-#################################################
-# Generic optimisation objective initialisation #
-#################################################
-
-function transform!!(f::OptimLogDensity)
-    ## Check link status of vi in OptimLogDensity
-    linked = DynamicPPL.istrans(f.varinfo)
+"""
+    link(p::ModeEstimationProblem)
 
-    ## transform into constrained or unconstrained space depending on current state of vi
-    f = Accessors.@set f.varinfo = if !linked
-        DynamicPPL.link!!(f.varinfo, f.model)
-    else
-        DynamicPPL.invlink!!(f.varinfo, f.model)
+Transform a mode estimation problem to unconstrained space, where all parameters can take
+any real value.
+"""
+function link(p::ModeEstimationProblem)
+    if has_constraints(p)
+        throw(
+            "Transforming constrained optimisation problems to unconstrained space is " *
+            "not yet implemented."
+        )
     end
-
-    return f
+    # Note that redefining obj and init_value out of place is intentional: It avoids
+    # issues with models for which linking changes the parameter space dimension.
+    ld = p.log_density
+    ld = Accessors.@set ld.varinfo = DynamicPPL.unflatten(ld.varinfo, copy(p.init_value))
+    ld = Accessors.@set ld.varinfo = DynamicPPL.link(ld.varinfo, ld.model)
+    init_value = ld.varinfo[:]
+    return ModeEstimationProblem(
+        p.model, p.estimator, init_value, p.constraints, ld, p.adtype, true
+    )
 end
 
-function transform!!(p::AbstractArray, vi::DynamicPPL.VarInfo, model::DynamicPPL.Model, ::constrained_space{true})
-    linked = DynamicPPL.istrans(vi)
-    
-    !linked && return identity(p)  # TODO: why do we do `identity` here?
-    vi = DynamicPPL.unflatten(vi, p)
-    vi = DynamicPPL.invlink!!(vi, model)
-    p .= vi[:]
-
-    # If linking mutated, we need to link once more.
-    linked && DynamicPPL.link!!(vi, model)
-
-    return p
-end
+"""
+    OptimizationProblem(me_prob::ModeEstimationProblem)
 
-function transform!!(p::AbstractArray, vi::DynamicPPL.VarInfo, model::DynamicPPL.Model, ::constrained_space{false})
-    linked = DynamicPPL.istrans(vi)
-    if linked
-        vi = DynamicPPL.invlink!!(vi, model)
+Create a SciML `OptimizationProblem` from a `ModeEstimationProblem`.
+"""
+function OptimizationProblem(me_prob::ModeEstimationProblem)
+    c = me_prob.constraints
```

Comment:
```suggestion
function OptimizationProblem(prob::ModeEstimationProblem)
    c = prob.constraints
```
IMO a bit nicer but no biggie :+1:

Reply:
Done